### PR TITLE
TASK: Case sensitive migration and migration of baseWorkspaceName

### LINKED ISSUE #5201: WorkspaceNames in legacy events may not meet new constraints
We need a migration to rewrite all workspaceNames in legacy events, to fulfill the new workspaceName constraints.

See: https://github.com/neos/neos-development-collection/issues/5125

Thanks to @bwaidelich:
> FYI: A migration could look something like
> 
> ```sql
> UPDATE
>   cr_default_events
> SET
>   payload = JSON_SET(payload, '$.workspaceName', CONCAT('w', MD5(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName')))))
> FROM
>   cr_default_events
> WHERE
>   JSON_EXTRACT(payload, '$.workspaceName') IS NOT NULL
>   AND JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName')) NOT REGEXP '^[a-z][a-z0-9\-]{0,30}$'
> ```

### PR DESCRIPTION
**Pull request by dlubitz**
Follow up of #5202 

- Increase workspaceName lenght to 36 chars.
- Lower constraints on content so also a hash or UUID is valid.
- Ensure migration works case sensitive. 
- Ensure newly created user-workspaceNames will be valid.
- Also applying to baseWorkspaceName.

Fixes #5201

### CODE REVIEW

**Comment by mhsdesign** [comment]
> Also applying to baseWorkspaceName.

ohhhh dammm yes i feel we are just getting started with the can of worms. Senf or mayo?

Just food for thought: Is it realistic that there will be one workspace `foo` already and another `Foo` and this will clash em? I guess not and i think we can ignore this case and for the case someone really has this problem help them via manual replace :D

**Comment by mhsdesign** [comment]
At first i thought it was really clever to do it via sql but maybe as its just a onetime thing we could go the manual php loop round? The thing is its getting more and more complex. I wouldnt mind this taking a little longer  ^^ Though i will also approve this and i admire that you can write sql like that :D

**Comment by dlubitz** [comment]
Should be fine now(TM) . We would also cover uppercase/lowercase by the md5 now.

File: Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php [hunk 1]
```diff
@@ -385,24 +385,35 @@ public function migratePayloadToValidWorkspaceNames(\Closure $outputFn): void
 
         $eventTableName = DoctrineEventStoreFactory::databaseTableName($this->contentRepositoryId);
         $this->connection->beginTransaction();
-        $statement = <<<SQL
+        $statementWorkspaceName = <<<SQL
                 UPDATE {$eventTableName}
                 SET
-                  payload = JSON_SET(payload, '$.workspaceName', SUBSTR(MD5(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName'))), 1, 30))
+                  payload = JSON_SET(payload, '$.workspaceName', LEFT(CONCAT('w', MD5(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName')))), 30))
```
Comment (bwaidelich) [inline Neos.ContentRepositoryRegistry/Classes/Service/EventMigrationService.php]: something like this should work, if we change the `WorkspaceName` regex to `^[a-z0-9\-]{0,36}`

```suggestion
                  payload = JSON_SET(
                      payload, 
                      '$.workspaceName',
                      IF(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName')) REGEXP '^[a-z0-9\-]{0,36}$',
                        LOWER(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName'))),
                        MD5(JSON_UNQUOTE(JSON_EXTRACT(payload, '$.workspaceName')))
                      )
                    )
```

File: Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php [hunk 2]
```diff
@@ -89,8 +89,7 @@ public static function transliterateFromString(string $name): self
 
         // If the name is still invalid at this point, we fall back to md5
         if (!self::hasValidFormat($name)) {
-            $prefix = 'workspace-';
-            $name = $prefix . substr(md5($originalName), 0, self::MAX_LENGTH - strlen($prefix));
+            $name = substr(md5($originalName), 0, self::MAX_LENGTH);
```
Comment (bwaidelich) [inline Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php:92]: Just curiously asking: why did you change that, I kinda liked the `workspace-` prefix?

**Review by bwaidelich — APPROVED** [review]
+1 by reading. Thank you! <3

File: Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php (hunk 2 above)
Comment (dlubitz) [reply Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php:92]: Because the md5 is 32 chars already. And so you would shorten it always.

Comment (bwaidelich) [reply Neos.ContentRepository.Core/Classes/SharedModel/Workspace/WorkspaceName.php:92]: ok, makes sense

**Review by nezaniel — CHANGES_REQUESTED** [review]
This still does not seem to work.... after applying the event migration I still get "Invalid workspace name (user-Nezaniel) given." during cr:projectionreplayall

**Comment by nezaniel** [comment]
I found the issue; There are two more workspaceName proeprties (sourceWorkspaceName and targetWorkspaceName). I fixed it and it now works on my starship(tm)

**Review by nezaniel — APPROVED** [review]

**Comment by mhsdesign** [comment]
post merge +1 thanks (and also thanks for the note with having to replay the workspace module ^^)
